builtInFunctions/setGuardian.go

```diff
@@ -0,0 +1,270 @@
+	default:
+		return &vmcommon.VMOutput{ReturnCode: vmcommon.ExecutionFailed}, nil
+	}
+}
+
+func (sg *setGuardian) checkArguments(
+	senderAccount,
+	receiverAccount vmcommon.UserAccountHandler,
+	vmInput *vmcommon.ContractCallInput,
+) error {
+	if check.IfNil(senderAccount) {
+		return fmt.Errorf("%w for sender", ErrNilUserAccount)
+	}
+	if check.IfNil(receiverAccount) {
+		return fmt.Errorf("%w for receiver", ErrNilUserAccount)
+	}
+	if vmInput == nil {
+		return ErrNilVmInput
+	}
+	if !(bytes.Equal(senderAccount.AddressBytes(), receiverAccount.AddressBytes()) &&
+		bytes.Equal(senderAccount.AddressBytes(), vmInput.CallerAddr)) {
+		return ErrOperationNotPermitted
+	}
+	if vmInput.CallValue == nil {
+		return ErrNilValue
+	}
+	if !isZero(vmInput.CallValue) {
+		return ErrBuiltInFunctionCalledWithValue
+	}
+	if len(vmInput.Arguments) != 1 {
+		return fmt.Errorf("%w, expected 1, got %d ", ErrInvalidNumberOfArguments, len(vmInput.Arguments))
+	}
+	if !sg.isAddressValid(vmInput.Arguments[0]) {
+		return fmt.Errorf("%w for guardian", ErrInvalidAddress)
+	}
+	if bytes.Equal(vmInput.CallerAddr, vmInput.Arguments[0]) {
+		return ErrCannotOwnAddressAsGuardian
+	}
+	if vmInput.GasProvided < sg.funcGasCost {
+		return ErrNotEnoughGas
+	}
+
+	return nil
+}
+
+func isZero(n *big.Int) bool {
+	return len(n.Bits()) == 0
+}
+
+func (sg *setGuardian) isAddressValid(addressBytes []byte) bool {
+	isLengthOk := len(addressBytes) == sg.pubKeyConverter.Len()
+	if !isLengthOk {
+		return false
+	}
+
+	encodedAddress := sg.pubKeyConverter.Encode(addressBytes)
+	return encodedAddress != ""
+}
+
+func (sg *setGuardian) guardians(account vmcommon.UserAccountHandler) (*Guardians, error) {
+	marshalledData, err := account.AccountDataHandler().RetrieveValue(sg.keyPrefix)
+	if err != nil {
+		return nil, err
+	}
+
+	// Fine, account has no guardian set
+	if len(marshalledData) == 0 {
+		return &Guardians{Data: make([]*Guardian, 0)}, nil
+	}
+
+	guardians := &Guardians{}
+	err = sg.marshaller.Unmarshal(guardians, marshalledData)
+	return guardians, err
+}
+
+func (sg *setGuardian) contains(guardians *Guardians, guardianAddress []byte) bool {
+	for _, guardian := range guardians.Data {
+		if bytes.Equal(guardian.Address, guardianAddress) {
+			return true
+		}
+	}
+	return false
+}
+
+func (sg *setGuardian) tryAddGuardian(
+	account vmcommon.UserAccountHandler,
+	guardianAddress []byte,
+	guardians *Guardians,
+	gasProvided uint64,
+) (*vmcommon.VMOutput, error) {
+	err := sg.addGuardian(account, guardianAddress, guardians)
+	if err != nil {
+		return nil, err
+	}
+	return &vmcommon.VMOutput{ReturnCode: vmcommon.Ok, GasRemaining: gasProvided - sg.funcGasCost}, nil
+}
+
+func (sg *setGuardian) addGuardian(account vmcommon.UserAccountHandler, guardianAddress []byte, guardians *Guardians) error {
+	guardian := &Guardian{
+		Address:         guardianAddress,
+		ActivationEpoch: sg.blockchainHook.CurrentEpoch() + sg.guardianActivationEpochs,
```

Comment:
Adding Guardian should be independent of epoch and guardian activation epoch. If all guardians plus the owner of the wallet signed to add a guardian, you should add a guardian. On remove you can done that same if all guardians signed remove guardian - you could do that.

Reply:
As discussed, we will keep the current structure as it is for now